android/app/src/main/java/com/zulipmobile/sharing/ReceiveShareActivity.kt

```diff
@@ -0,0 +1,98 @@
+import com.facebook.react.ReactActivity
+import com.facebook.react.ReactApplication
+import com.facebook.react.bridge.ReactContext
+import com.facebook.react.bridge.WritableMap
+import com.facebook.react.bridge.Arguments
+import com.zulipmobile.notifications.*
+
+const val TAG = "ZulipReceiveShare"
+
+class ReceiveShareActivity : ReactActivity() {
+
+    /**
+     * Returns the name of the main component registered from JavaScript.
+     * This is used to schedule rendering of the component.
+     */
+    override fun getMainComponentName(): String? {
+        return "SharingRoot"
+    }
+
+    private fun sendEvent(reactContext: ReactContext,
+                          eventName: String,
+                          @Nullable params: WritableMap) {
+        Log.d(TAG, "Sending event with shared data")
+        emit(reactContext, eventName, params)
+    }
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        WebView.setWebContentsDebuggingEnabled(true)
+        if (intent?.action == Intent.ACTION_SEND) {
+            handleSend(intent)
+        }
+        finish()
+    }
+
+    private fun handleSend(intent: Intent) {
+        val application = application as ReactApplication
+        val host = application.reactNativeHost
+        val reactContext = host.tryGetReactInstanceManager()?.currentReactContext
+
+        val params = (if ("text/plain" == intent.type) {
+            getParamsFromTextIntent(intent)
+        } else if (intent.type?.startsWith("image/") == true) {
+            getParamsFromImageIntent(intent)
+        } else {
+            getParamsFromFileIntent(intent)
+        })
+            ?: throw Exception("Could not parse received sharing data")
+
+        if (null == reactContext) {
+            launchMainActivity(application as Context)
+            SharingModule.initialSharedData = params
+        } else {
+            sendEvent(reactContext, "shareReceived", params)
+        }
+    }
+
+    private fun getParamsFromTextIntent(intent: Intent): WritableMap {
+        val sharedText = intent.getStringExtra(Intent.EXTRA_TEXT)
+        val params = Arguments.createMap()
+        params.putString("type", "text")
+        params.putString("sharedText", sharedText)
+        return params
+    }
+
+    private fun getParamsFromImageIntent(intent: Intent): WritableMap? {
+        val uri = intent.getParcelableExtra(Intent.EXTRA_STREAM) as? Uri ?: return null
+        val params = Arguments.createMap()
+        params.putString("type", "image")
+        params.putString("sharedImageUri", uri.toString())
+        return params
+    }
+
+    private fun getParamsFromFileIntent(intent: Intent): WritableMap? {
+        val uri = intent.getParcelableExtra(Intent.EXTRA_STREAM) as? Uri ?: return null
+        val params = Arguments.createMap()
+        params.putString("type", "file")
+        params.putString("sharedFileUri", uri.toString())
+        return params
+    }
```

Comment:
I think we can deduplicate some more of this by having a method like `getParamsFromIntent` -- which contains the `"text/plain" == intent.type` etc. logic, and also contains the contents of these three methods.

That way the `params = Arguments.createMap()`, and the couple of lines of overhead for making a separate method, get shared across the three cases. Also the new method has a bit more clear and self-contained of an interface than each of these three.

Then `handleSend` would just call that method, like `val params = getParamsFromIntent(intent)`.